observ/store.py

```diff
@@ -0,0 +1,135 @@
+import copy
+from functools import partial, wraps
+from inspect import signature
+from typing import Callable, TypeVar
+
+from .dep import Dep
+from .observables import (
+    reactive,
+    readonly,
+    shallow_reactive,
+    to_raw,
+)
+from .watcher import Watcher
+
+
+T = TypeVar("T", bound=Callable)
+
+
+registry = {}
+
+
+def mutation(fn: T) -> T:
+    registry[fn] = "mutation"
+
+    @wraps(fn)
+    def inner(self, *args, **kwargs):
+        return fn(self, *args, **kwargs)
+
+    return inner
+
+
+def computed(fn: T) -> T:
+    parameters = signature(fn).parameters
+    if len(parameters) == 1 and "self" in parameters:
+        registry[fn] = "computed"
+
+        @wraps(fn)
+        def inner(self):
+            return fn(self)
+
+        return inner
+    else:
+        watcher = Watcher(fn)
+
+        @wraps(fn)
+        def getter():
+            if watcher.dirty:
+                watcher.evaluate()
+            if Dep.stack:
+                watcher.depend()
+            return watcher.value
+
+        getter.__watcher__ = watcher
+        return getter
```

Comment:
This is confusing, why not just `return computed(fn)`?

Reply:
Because `computed` *is* now this function. A bit confusing, yes. I might split it up again, but then I'll probably have to rename one of them.

Reply:
That.. is too complicated :) let's think of something to make it nicer

Reply:
Ok, I came up with the following: if we don't expose the `store` through the api, then there doesn't need to be a name clash. But in order to prevent circular imports I've moved the original definition of `computed` to [/observ/watcher.py]()

Because the store is a pretty specific piece of code, I think it would be okay to leave it out of the default api. If someone wants to implement a subclass of the `Store`, then the following imports would suffice:
```python
from observ.store import computed, mutation, Store
```